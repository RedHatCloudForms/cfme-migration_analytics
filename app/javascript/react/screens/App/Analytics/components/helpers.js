export const readFirstFile = (fileHandles, onLoad) => {
  if (fileHandles && fileHandles.length > 0) {
    const reader = new FileReader();
    reader.onload = () => onLoad({ filename: fileHandles[0].name, body: reader.result });
    reader.readAsBinaryString(fileHandles[0]);
  }
};
