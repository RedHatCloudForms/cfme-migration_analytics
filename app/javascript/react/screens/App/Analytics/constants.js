export const PROVIDERS_SUMMARY_REPORT_FILTERS = {
  rpt_group: 'Migration Readiness - Providers',
  name: 'VMware Environment Summary'
};
