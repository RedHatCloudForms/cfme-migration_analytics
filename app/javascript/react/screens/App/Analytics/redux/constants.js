export const FETCH_MANIFEST_INFO = 'MA_FETCH_MANIFEST_INFO';
export const TOGGLE_MANIFEST_UPDATE_MODAL = 'MA_TOGGLE_MANIFEST_UPDATE_MODAL';
export const CALCULATE_SUMMARY_DATA = 'MA_CALCULATE_SUMMARY_DATA';
export const SELECT_PROVIDERS = 'MA_SELECT_PROVIDERS';
export const SELECT_DETAILED_DATA = 'MA_SELECT_DETAILED_DATA';
export const START_INVENTORY_BUNDLE = 'MA_START_INVENTORY_BUNDLE';
export const FETCH_BUNDLE_TASK = 'MA_FETCH_BUNDLE_TASK';
export const RESET_DATA_COLLECTION_STATE = 'MA_RESET_DATA_COLLECTION_STATE';

export const MANIFEST_INFO_URL = '/api/red_hat_migration_analytics';
export const INVENTORY_BUNDLE_URL = '/api/red_hat_migration_analytics';
