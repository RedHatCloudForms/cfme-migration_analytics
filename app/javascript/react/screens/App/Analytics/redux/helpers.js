export const calculateSummaryData = ({ vmSummaryReportResult, envSummaryReportResult }) => {
  const valuesByProviderId = {};

  vmSummaryReportResult.result_set.forEach(vm => {
    const {
      'ext_management_system.id': providerId,
      'ext_management_system.name': providerName,
      allocated_disk_storage: allocatedDiskSpace,
      mem_cpu: allocatedMemory,
      cpu_total_cores: numCpuCores
    } = vm;
    if (!valuesByProviderId[providerId]) {
      valuesByProviderId[providerId] = {
        id: providerId,
        name: providerName,
        numHypervisors: 0,
        numVms: 0,
        allocatedDiskSpace: 0,
        allocatedMemory: 0,
        numCpuCores: 0
      };
    }
    const provider = valuesByProviderId[providerId];
    provider.numVms += 1;
    provider.allocatedDiskSpace += allocatedDiskSpace;
    provider.allocatedMemory += allocatedMemory;
    provider.numCpuCores += numCpuCores;
  });

  envSummaryReportResult.result_set.forEach(hypervisor => {
    const providerId = hypervisor['ext_management_system.id'];
    valuesByProviderId[providerId].numHypervisors += 1;
  });

  const total = {
    numProviders: Object.keys(valuesByProviderId).length,
    numHypervisors: 0,
    numVms: 0,
    allocatedDiskSpace: 0,
    allocatedMemory: 0,
    numCpuCores: 0
  };

  Object.keys(valuesByProviderId).forEach(providerId => {
    const provider = valuesByProviderId[providerId];
    total.numHypervisors += provider.numHypervisors;
    total.numVms += provider.numVms;
    total.allocatedDiskSpace += provider.allocatedDiskSpace;
    total.allocatedMemory += provider.allocatedMemory;
    total.numCpuCores += provider.numCpuCores;
  });

  return {
    total,
    providers: Object.values(valuesByProviderId)
  };
};
