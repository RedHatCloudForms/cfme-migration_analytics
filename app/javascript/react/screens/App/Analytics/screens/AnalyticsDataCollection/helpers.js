const payloadPathRegex = /\/migration_analytics\/.+$/;

export const getPayloadUrl = payloadPath => {
  const match = payloadPath && payloadPath.match(payloadPathRegex);
  if (!match) return null;
  const { protocol, hostname, port } = window.location;
  const origin = `${protocol}//${hostname}${port ? `:${port}` : ''}`;
  return `${origin}${match[0]}`;
};
