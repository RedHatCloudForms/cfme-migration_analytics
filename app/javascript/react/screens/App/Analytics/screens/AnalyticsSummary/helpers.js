export const someProvidersExist = ({ providers }) => providers && providers.length > 0;

export const noRefreshErrorsExist = ({ providersWithRefreshErrors }) =>
  providersWithRefreshErrors && providersWithRefreshErrors.length === 0;

export const someProvidersAwaitingRefresh = ({ providersAwaitingRefresh }) =>
  providersAwaitingRefresh && providersAwaitingRefresh.length > 0;

export const noProvidersAwaitingRefresh = ({ providersAwaitingRefresh }) =>
  providersAwaitingRefresh && providersAwaitingRefresh.length === 0;

export const providersRefreshed = props =>
  someProvidersExist(props) && noRefreshErrorsExist(props) && noProvidersAwaitingRefresh(props);
