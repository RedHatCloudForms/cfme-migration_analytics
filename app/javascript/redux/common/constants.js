export const RESET_ALL_STATE = 'MA_RESET_ALL_STATE';
