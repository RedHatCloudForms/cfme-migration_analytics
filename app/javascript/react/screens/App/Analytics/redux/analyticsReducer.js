import Immutable from 'seamless-immutable';
import {
  CALCULATE_SUMMARY_DATA,
  SELECT_PROVIDERS,
  SELECT_DETAILED_DATA,
  FETCH_MANIFEST_INFO,
  TOGGLE_MANIFEST_UPDATE_MODAL,
  START_INVENTORY_BUNDLE,
  FETCH_BUNDLE_TASK,
  RESET_DATA_COLLECTION_STATE,
  CHANGE_MANIFEST
} from './constants';
import { functionLookupReducer, getHandlersForBasicFetchActions } from '../../../../../redux/helpers';
import { calculateSummaryData } from './helpers';
import { RESET_ALL_STATE } from '../../../../../redux/common/constants';

const fetchManifestInfo = getHandlersForBasicFetchActions(
  FETCH_MANIFEST_INFO,
  'isFetchingManifestInfo',
  'errorFetchingManifestInfo',
  'manifestInfo'
);

const changeManifest = getHandlersForBasicFetchActions(
  CHANGE_MANIFEST,
  'isChangingManifest',
  'errorChangingManifest',
  'changeManifestResult'
);

const startInventoryBundle = getHandlersForBasicFetchActions(
  START_INVENTORY_BUNDLE,
  'isStartingBundle',
  'errorStartingBundle',
  'bundleStartResult'
);

const fetchBundleTask = getHandlersForBasicFetchActions(
  FETCH_BUNDLE_TASK,
  'isFetchingBundleTask',
  'errorFetchingBundleTask',
  'bundleTask'
);

export const initialState = Immutable({
  ...fetchManifestInfo.initialState,
  ...changeManifest.initialState,
  manifestUpdateModalVisible: false,
  summaryData: null,
  selectedProviders: [],
  detailedDataSelected: false,
  ...startInventoryBundle.initialState,
  ...fetchBundleTask.initialState
});

const actionHandlers = {
  ...fetchManifestInfo.actionHandlers,
  ...changeManifest.actionHandlers,
  [TOGGLE_MANIFEST_UPDATE_MODAL]: (state, action) =>
    state.set('manifestUpdateModalVisible', action.show !== null ? action.show : !state.manifestUpdateModalVisible),
  [CALCULATE_SUMMARY_DATA]: (state, action) => state.set('summaryData', calculateSummaryData(action.results)),
  [SELECT_PROVIDERS]: (state, action) => state.set('selectedProviders', action.selectedProviders),
  [SELECT_DETAILED_DATA]: (state, action) => state.set('detailedDataSelected', action.detailedDataSelected),
  ...startInventoryBundle.actionHandlers,
  ...fetchBundleTask.actionHandlers,
  [RESET_DATA_COLLECTION_STATE]: state =>
    Immutable.merge(state, {
      ...startInventoryBundle.initialState,
      ...fetchBundleTask.initialState
    }),
  [RESET_ALL_STATE]: () => initialState
};

export default functionLookupReducer(initialState, actionHandlers);
