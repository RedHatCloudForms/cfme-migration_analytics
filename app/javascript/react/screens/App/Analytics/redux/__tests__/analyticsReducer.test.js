import analyticsReducer, { initialState } from '../analyticsReducer';
import {
  CALCULATE_SUMMARY_DATA,
  SELECT_PROVIDERS,
  SELECT_DETAILED_DATA,
  RESET_DATA_COLLECTION_STATE,
  FETCH_MANIFEST_INFO,
  START_INVENTORY_BUNDLE,
  FETCH_BUNDLE_TASK
} from '../constants';
import { calculateSummaryData } from '../helpers';
import { RESET_ALL_STATE } from '../../../../../../redux/common/constants';

describe('analytics reducer', () => {
  test('constructs initial state correctly', () => {
    expect(initialState).toMatchSnapshot();
  });

  test('calls an action handler for fetchManifestInfo', () => {
    const state = analyticsReducer(initialState, { type: `${FETCH_MANIFEST_INFO}_PENDING` });
    expect(state.isFetchingManifestInfo).toBe(true);
  });

  test('calculate summary data', () => {
    const action = {
      type: CALCULATE_SUMMARY_DATA,
      results: {
        vmSummaryReportResult: { result_set: [] },
        envSummaryReportResult: { result_set: [] }
      }
    };
    expect(analyticsReducer(initialState, action).summaryData).toEqual(calculateSummaryData(action.results));
  });

  test('select providers', () => {
    const action = {
      type: SELECT_PROVIDERS,
      selectedProviders: ['dummy', 'array']
    };
    expect(analyticsReducer(initialState, action).selectedProviders).toEqual(action.selectedProviders);
  });

  test('select detailed data', () => {
    const action = {
      type: SELECT_DETAILED_DATA,
      detailedDataSelected: true
    };
    expect(analyticsReducer(initialState, action).detailedDataSelected).toBe(true);
  });

  test('calls an action handler for startInventoryBundle', () => {
    const state = analyticsReducer(initialState, { type: `${START_INVENTORY_BUNDLE}_PENDING` });
    expect(state.isStartingBundle).toBe(true);
  });

  test('calls an action handler for fetchBundleTask', () => {
    const state = analyticsReducer(initialState, { type: `${FETCH_BUNDLE_TASK}_PENDING` });
    expect(state.isFetchingBundleTask).toBe(true);
  });

  test('reset data collection state', () => {
    const prevState = {
      ...initialState,
      isStartingBundle: true,
      isFetchingBundleTask: true
    };
    const state = analyticsReducer(prevState, {
      type: RESET_DATA_COLLECTION_STATE
    });
    expect(state).toEqual(initialState);
  });

  test('reset all state', () => {
    const prevState = {
      ...initialState,
      summaryData: { dummy: 'object' },
      selectedProviders: ['dummy', 'array'],
      detailedDataSelected: true
    };
    const state = analyticsReducer(prevState, {
      type: RESET_ALL_STATE
    });
    expect(state).toEqual(initialState);
  });
});
