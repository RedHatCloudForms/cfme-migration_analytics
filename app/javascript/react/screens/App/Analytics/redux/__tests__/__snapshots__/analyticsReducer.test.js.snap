// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analytics reducer constructs initial state correctly 1`] = `
Object {
  "bundleStartResult": null,
  "bundleTask": null,
  "changeManifestResult": null,
  "detailedDataSelected": false,
  "errorChangingManifest": null,
  "errorDownloadingPayload": null,
  "errorFetchingBundleTask": null,
  "errorFetchingManifestInfo": null,
  "errorStartingBundle": null,
  "isChangingManifest": false,
  "isDownloadingPayload": false,
  "isFetchingBundleTask": false,
  "isFetchingManifestInfo": false,
  "isStartingBundle": false,
  "manifestInfo": null,
  "manifestUpdateModalVisible": false,
  "payloadFileName": null,
  "selectedProviders": Array [],
  "summaryData": null,
}
`;
